File: plugins/plugin-roblox/typescript/actions/sendGameMessage.ts
import {
  type Action,
  type ActionExample,
  type ActionResult,
  type HandlerCallback,
  type HandlerOptions,
  type IAgentRuntime,
  logger,
  type Memory,
  type State,
} from "@elizaos/core";
import type { RobloxService } from "../services/RobloxService";
import { ROBLOX_SERVICE_NAME } from "../types";

const actionName = "SEND_ROBLOX_MESSAGE";

const sendGameMessageExamples: ActionExample[][] = [
  [
    {
      name: actionName,
      content: {
        text: "Tell everyone in the game that there's a special event happening",
      },
    },
    {
      name: actionName,
      content: {
        text: "I'll announce the special event to all players in the game!",
        action: "SEND_ROBLOX_MESSAGE",
      },
    },
  ],
  [
    {
      name: "{{user1}}",
      content: {
        text: "Send a message to player123 welcoming them to the game",
      },
    },
    {
      name: "{{agentName}}",
      content: {
        text: "I'll send a personalized welcome message to player123.",
        action: "SEND_ROBLOX_MESSAGE",
      },
    },
  ],
];

const sendGameMessage: Action = {
  name: actionName,
  similes: [],
  description: "Send a message to players in the Roblox game.",
  examples: sendGameMessageExamples,

  validate: async (runtime: IAgentRuntime): Promise<boolean> => {
    const apiKey = runtime.getSetting("ROBLOX_API_KEY");
    const universeId = runtime.getSetting("ROBLOX_UNIVERSE_ID");
    return Boolean(apiKey && universeId);
  },

  handler: async (
    runtime: IAgentRuntime,
    message: Memory,
    state: State | undefined,
    _options?: HandlerOptions,
    callback?: HandlerCallback
  ): Promise<ActionResult | undefined> => {
    try {
      const service = runtime.getService<RobloxService>(ROBLOX_SERVICE_NAME);
      if (!service) {
        logger.error("Roblox service not found");
        if (callback) {
          callback({
            text: "Roblox service not available.",
            action: "SEND_ROBLOX_MESSAGE",
          });
        }
        return {
          success: false,
          error: "Roblox service not found",
        };
      }

      const messageContent =
        (state?.message as string) || (message.content as { text?: string }).text || "";

      if (!messageContent) {
        logger.warn("No message content to send");
        if (callback) {
          callback({
            text: "I need a message to send to the game.",
            action: "SEND_ROBLOX_MESSAGE",
          });
        }
        return {
          success: false,
          error: "No message content to send",
        };
      }

      const targetPlayerIds = extractPlayerIds(messageContent);

      await service.sendMessage(runtime.agentId, messageContent, targetPlayerIds);

      logger.info(
        { targetPlayerIds, messageLength: messageContent.length },
        "Sent message to Roblox game"
      );

      if (callback) {
        const targetInfo =
          targetPlayerIds && targetPlayerIds.length > 0
            ? `to ${targetPlayerIds.length} specific player(s)`
            : "to all players";
        callback({
          text: `I've sent the message ${targetInfo} in the game.`,
          action: "SEND_ROBLOX_MESSAGE",
        });
      }

      return {
        success: true,
        text: `Sent message ${targetPlayerIds && targetPlayerIds.length > 0 ? `to ${targetPlayerIds.length} specific player(s)` : "to all players"} in the game`,
        data: {
          targetPlayerIds,
          messageLength: messageContent.length,
        },
      };
    } catch (error) {
      logger.error({ error }, "Failed to send Roblox message");
      const errorMessage = error instanceof Error ? error.message : "Unknown error";
      if (callback) {
        callback({
          text: "Error sending message.",
          action: "SEND_ROBLOX_MESSAGE",
        });
      }
      return {
        success: false,
        error: errorMessage,
      };
    }
  },
};

function extractPlayerIds(message: string): number[] | undefined {
  const playerIdPattern = /\bplayer\s*(\d+)\b/gi;
  const matches = [...message.matchAll(playerIdPattern)];

  if (matches.length > 0) {
    return matches.map((m) => parseInt(m[1], 10));
  }

  return undefined;
}

export default sendGameMessage;

File: plugins/plugin-roblox/python/elizaos_plugin_roblox/actions/send_game_message.py
from __future__ import annotations

from dataclasses import dataclass


@dataclass
class ActionExample:
    input: str
    output: str


@dataclass
class SendGameMessageAction:
    @property
    def name(self) -> str:
        return "SEND_ROBLOX_MESSAGE"

    @property
    def similes(self) -> list[str]:
        return [
            "ROBLOX_MESSAGE",
            "GAME_MESSAGE",
            "SEND_TO_GAME",
            "BROADCAST_MESSAGE",
            "TELL_PLAYERS",
        ]

    @property
    def description(self) -> str:
        return "Send a message to players in a Roblox game. Can target all players or specific player IDs."

    def _is_send_message_request(self, text: str) -> bool:
        lower = text.lower()
        has_action = any(word in lower for word in ["send", "tell", "message", "broadcast"])
        has_target = any(word in lower for word in ["game", "player", "roblox"])
        return has_action and has_target

    async def validate(self, message_text: str) -> bool:
        return self._is_send_message_request(message_text)

    async def handler(self, params: dict[str, object]) -> dict[str, object]:
        content = params.get("content")
        if not content or not isinstance(content, str):
            raise ValueError("Missing 'content' parameter")

        target_player_ids = params.get("target_player_ids")

        return {
            "action": "SEND_ROBLOX_MESSAGE",
            "content": content,
            "target_player_ids": target_player_ids,
            "status": "pending",
        }

    @property
    def examples(self) -> list[ActionExample]:
        return [
            ActionExample(
                input="Tell everyone in the game that there's a special event happening",
                output="I'll announce the special event to all players in the game!",
            ),
            ActionExample(
                input="Send a message to player123 welcoming them to the game",
                output="I'll send a personalized welcome message to player123.",
            ),
        ]
